File: 0848_shifting_letters/shifting_letters.py
from typing import List


def shifting_letters(s: str, shifts: List[int]) -> str:
    """
    Time: O(n)
    Space: O(n)
    """
    res = ""
    for i in range(len(s) - 2, -1, -1):
        shifts[i] = shifts[i] + shifts[i + 1]
    for i, c in enumerate(s):
        res += shift_char_by_indexes(c, shifts[i])
    return res
    
def shift_char_by_indexes(char: str, num_of_indexes: int) -> str:
    a = ord('a')
    return chr((ord(char) - a + num_of_indexes) % 26 + a)

File: 0848_shifting_letters/shifting_letters.java
class Solution {
    // Time: O(n)
    // Space: O(n)
    public String shiftingLetters(String s, int[] shifts) {
        StringBuilder res = new StringBuilder(s);
        for (int i = shifts.length - 2; i >= 0; i--) {
            shifts[i] = (shifts[i] + shifts[i + 1]) % 26;
        }
        for (int i = 0; i < s.length(); i++) {
            char shiftedChar = shiftCharByIndexes(s.charAt(i), shifts[i]);
            res.setCharAt(i, shiftedChar);
        }
        return res.toString();
    }
    
    private char shiftCharByIndexes(char c, int numOfIndexes) {
        return (char)((c - 'a' + numOfIndexes) % 26 + 'a');
    }
}